src/emqx_broker.erl

```diff
@@ -389,33 +387,38 @@ code_change(_OldVsn, State, _Extra) ->
 %% Internal functions
 %%------------------------------------------------------------------------------
 
-do_subscribe(Group, Topic, Subscriber, SubOpts) ->
+do_subscribe(undefined, Topic, Subscriber, SubOpts) ->
+    ets:insert(?SUBSCRIPTION, {Subscriber, Topic}),
+    ets:insert(?SUBSCRIBER, {Topic, Subscriber}),
+    ets:insert(?SUBOPTION, {{Topic, Subscriber}, SubOpts});
+
+do_subscribe(Group, Topic, Subscriber  = {SubPid, _}, SubOpts) ->
+    emqx_shared_sub:subscribe(Group, Topic, SubPid),
     ets:insert(?SUBSCRIPTION, {Subscriber, shared(Group, Topic)}),
-    ets:insert(?SUBSCRIBER, {Topic, shared(Group, Subscriber)}),
     ets:insert(?SUBOPTION, {{Topic, Subscriber}, SubOpts}).
 
-do_unsubscribe(Group, Topic, Subscriber) ->
+do_unsubscribe(undefined, Topic, Subscriber) ->
+    ets:delete_object(?SUBSCRIPTION, {Subscriber, Topic}),
+    ets:delete_object(?SUBSCRIBER, {Topic, Subscriber}),
+    ets:delete(?SUBOPTION, {Topic, Subscriber});
+
+do_unsubscribe(Group, Topic, Subscriber = {SubPid, _}) ->
+    emqx_shared_sub:unsubscribe(Group, Topic, SubPid),
```

Comment:
I noticed that `emqx_shared_sub:unsubscribe` only deletes the record from mnesia.
Please update it to also cast a message to `emqx_shared_sub` so it can demonitor the subscriber pid

Reply:
Maybe this is not necessary if the only case to unsubscribe is to shutdown (or kill) the subscriber pid.
then it would make more sense to rename the API from `unsubscribe` to `subscriber_down`